packages/ptokens-assets-evm/src/ptokens-evm-asset-builder.ts

```diff
@@ -30,25 +30,21 @@ export class pTokensEvmAssetBuilder extends pTokensAssetBuilder {
       method: 'hub',
       abi: factoryAbi,
     })
-
-    if (this._assetInfo.assetTokenAddress !== this._assetInfo.underlyingAssetTokenAddress) {
-      const pTokenAddress = await this._provider.makeContractCall<string, [string, string, number, string, string]>(
-        {
-          contractAddress: this.factoryAddress,
-          method: 'getPTokenAddress',
-          abi: factoryAbi,
-        },
-        [
-          this.assetInfo.underlyingAssetName,
-          this.assetInfo.underlyingAssetSymbol,
-          this.assetInfo.underlyingAssetDecimals,
-          this.assetInfo.underlyingAssetTokenAddress,
-          this.assetInfo.underlyingAssetNetworkId,
-        ],
-      )
-      if (!this.assetInfo.assetTokenAddress) this.assetInfo.assetTokenAddress = pTokenAddress
-      else if (pTokenAddress !== this.assetInfo.assetTokenAddress) throw new Error('Invalid pToken address')
-    }
+    const pTokenAddress = await this._provider.makeContractCall<string, [string, string, number, string, string]>(
+      {
+        contractAddress: this.factoryAddress,
+        method: 'getPTokenAddress',
+        abi: factoryAbi,
+      },
+      [
+        this.assetInfo.underlyingAssetName,
+        this.assetInfo.underlyingAssetSymbol,
+        this.assetInfo.underlyingAssetDecimals,
+        this.assetInfo.underlyingAssetTokenAddress,
+        this.assetInfo.underlyingAssetNetworkId,
+      ],
+    )
+    if (!this.assetInfo.assetTokenAddress) this.assetInfo.assetTokenAddress = pTokenAddress
```

Comment:
```suggestion
    if (!this.assetInfo.assetTokenAddress || this.assetInfo.assetTokenAddress === this.assetInfo.underlyingAddress ) this.assetInfo.assetTokenAddress = pTokenAddress
```
In this way you can remove pTokenAddress from the builder attributes.

Reply:
Thinking again all the thing.
I would add a isNative and pTokenAddress field in assetInfo, not necessarily required.
Then, in the builder you check for consistency:
- if isNative and assetTokenAddress != underlyingAddress throw
- if assetTokenAddress == underlyingAddress isNative = true
- if assetTokenAddress != underlyingAddress && assetTokenAddress != pTokenAddress throw
- if !assetTokenAddress { isNative = true, assign pTokenAddress }
something like that

Reply:
Then, in the swap function, you pass values depending on isNative.

Reply:
Not implemented yet. Current implementation follows directly https://github.com/pnetwork-association/pnetwork/blob/master/packages/ptokens-evm-contracts/contracts/core/PNetworkHub.sol#L585
isNative can improve usability. Implementation ongoing together with refactor below.